Remove the rewrite(exp) from solve()

asmeurer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes https://github.com/sympy/sympy/issues/24368
Alternative to #24549

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- solvers
  - Fix recursion error issues with solve().
<!-- END RELEASE NOTES -->

asmeurer committed:
Handle exponentials without matching bases better in _invert

This will make it possible to remove the generic rewrite(exp) from solve().

asmeurer committed:
Don't do rewrite(exp) in solve()

This can lead to recursion errors (e.g., #24368) because it is too easy for
the expression to get rewritten back to the way it was again in a recursive
call.

This still has some test failures. We need to handle trigonometric equations
in some other way (e.g., by doing a more targeted rewrite of them).

asmeurer commented:
This is the first part of fixing the issues as described at https://github.com/sympy/sympy/issues/24368#issuecomment-1397537927. I'm not sure yet how we want to handle trig functions. I pushed up a commit removing the rewrite so you can see the test failures.

Removing the rewrite(exp) also causes this test to fail:

```py
    assert solve([-y*(x + y - 1)/2, (y - 1)/x/y + 1/y],
        set=True, check=False) == ([x, y], {(1 - y, y), (x, 0)})
```

The reason is that the rewrite(exp) rewrites `1/y` as `exp(-log(y))`, which causes it to take a different solve path. 

Without it, it returns `([x, y], set())`. The `(x, 0)` solution doesn't actually pass through when `check=True`. I'm not sure if we really need to require that a certain solution be returned with `check=False` that would be omitted by default. I don't know about the `(1 - y, y)` solution. With check=True, it is still returned (but in a different form), but not when check=False:

```py
>>> a, b = -y*(x + y - 1)/2, (y - 1)/x/y + 1/y
>>> solve([a, b], set=True)
([x], {(1 - y,)})
>>> solve([a, b], set=True, check=False)
([x, y], set())
```

I'm not sure what the exact semantics are supposed to be with `set=True` and/or `check=False`. Is this new output above OK to just replace in the test, or should we investigate how to get the old behavior to continue to work? 

@smichr, this does suggest to me that the rewrite(exp) is in fact doing something for non-transcendental functions. I don't know if it ever matters in cases where the solution isn't just filtered by check=False. That will require some more playing around (my guess is no, since the polynomial solvers should be pretty good anyway, but I'm not confident about that).

asmeurer committed:
Re-enable the exp rewrite in solve but just for trig functions

oscarbenjamin commented on sympy/solvers/solvers.py:
```diff
@@ -2747,7 +2747,7 @@ def equal(expr1, expr2):
-        rewrite = lhs.rewrite(exp)
+        rewrite = lhs.rewrite([TrigonometricFunction, HyperbolicFunction], exp, evaluate=True)
         if rewrite != lhs:
             return _vsolve(rewrite - rhs, sym, **flags)
```
This approach of recursing if rewrite changes something is inherently fragile. If you trace through the calls from `_vsolve` back to here how can you ensure that passing `rewrite - lhs` will not just lead to this function receiving the same `eq` that it started with?

asmeurer replied:
I agree. This change makes it slightly less fragile (and fixes the one known issue), but if anything ever rewrites the trig functions back to exponentials, we're back where we started. Ideally solve would either not be written recursively at all, or be written in such a way that infinite loops/recursions are impossible. 

I think solve (or solveset) should be structured in a way that is much more explicit about the steps involved for solving an equation. There are three core steps involved in solving an equation

- Matching: similar to the matching step in the ODE solvers. Trying to figure out from the input which of many solvers should be attempted. This can also just be considered an optimization since you can just try the solvers in order til you find one that works, assuming they can never produce a wrong answer on invalid input. 
- Rewriting: trying various rewrites to put an expression into a form fit for a given solver. Only some rewrites should be tried for a given equation and for a given solver. Simplification is also a form of rewriting.
- Decomposition: writing f(x) as g(h(x)) so that g and h can be inverted separately. 

And then at the core there are specific "inversion" solvers, similar to the ODE solvers (e.g., a solver that can invert polynomials, a solver that knows how to produce LambertW-type solutions, etc.).

These should be split out into separate routines so that a central algorithm can apply them. dsolve is a decent model of how this sort of thing can be done. dsolve() can solve way more "types" of equations than solve(), and it's clear that that sort of approach can scale (although dsolve doesn't yet handle rewriting and decomposition very well). 

Right now matching is done in an ad hoc manner, and rewriting and recursion are done naively using recursion. But that means that everything is muddled together. It also means that there's no central algorithm that knows what is going on. That's why you get recursion errors, because each step only knows its own step of the process, and doesn't know about what has already been tried. 

solveset is a little cleaner in how its written, but it still doesn't really use this design.

oscarbenjamin replied:
It's okay to use recursion provided you know that you are recursing onto something that will not loop back so it should be e.g. something that is smaller. There is nothing wrong with recursing say from an expression to call an operation on its args because they are necessarily "smaller" and therefore cannot loop back.

The rewrite here is okay but it should be done before entering the recursive part of solve. That's what happens with hyperbolic functions:
https://github.com/sympy/sympy/blob/68803e50d9d76a8b14eaf2a537f90cb86d0a87be/sympy/solvers/solvers.py#L965-L968

The problem here is that rewrite produces something equivalent to what we started with and the rest of the solve code deliberately tries to canonicalise equivalent things. If you condense the recursion here to a single function it looks basically like this:
```python
def _solve(eq_orig):
    # transform all equivalent equations to a canonical form
    eq_canon = canonicalise(eq_orig)

   # do lots of stuff and return if it works.
   # if we get to the lines below then the stuff didn't work

   # change to a different but equivalent equation
   eq_changed = eq_canon.rewrite()
   if eq_changed != eq_canon:
       return _solve(eq_changed)
```
It's obvious that this can lead to infinite recursion which is why we have `if eq_changed != eq_canon` but that is not a robust guard. The *minimal* check to guard against infinite recursion should be `if canonicalise(eq_changed) != eq_canon` but there's no way to check that because there is no callable `canonicalise` function (it's spread out over all of `solve`).

asmeurer replied:
I'm not sure if there is a suggestion for a change here. 

The problem with your general idea is you cannot extract `canonicalize()` into a single operation. It happens all over the codebase, not just in solve. For this particular issue, the canonicalization happens in Poly. But it really could happen anywhere because it's the result of rewriting an unevaluated expression back into its evaluated form. The motivation in #24549 is to not do rewrites that produce unevaluated expressions, so at least that level of basic re-canonicalization won't break this logic.

oscarbenjamin replied:
> The problem with your general idea is you cannot extract `canonicalize()` into a single operation

I wan't proposing that as a solution but just as a demonstration of why the implicit assumption in this recursive call is broken.

To be clear I don't object to the change highlighted in the diff above which can reduce the problem in some cases. I just don't consider it to be a proper fix because the root problem with this code is still there.

asmeurer replied:
I think it would also be possible to a more complex recursion error where rewrites cycle back to each other. The only real way to fix this is to have a top-level solve algorithm that applies the rewrites and other heuristics and keeps track of what has been tried. And that design has other advantages too like making the algorithms more modular.

asmeurer committed:
Remove unused flag

This flag was relevant for a different PR.

asmeurer commented:
Assuming we're OK with the `rewrite([TrigonometricFunction, HyperbolicFunction], exp)` workaround, the only blocker here is whether the changed test output of `solve([-y*(x + y - 1)/2, (y - 1)/x/y + 1/y], set=True, check=False)` is acceptable (see https://github.com/sympy/sympy/pull/24553#issuecomment-1397707260).

oscarbenjamin commented:
> I'm not sure if we really need to require that a certain solution be returned with `check=False` that would be omitted by default.

We shouldn't worry about that.

> I don't know about the `(1 - y, y)` solution

It is a correct solution set for almost all possible values of y:
```python
In [11]: eqs = [-y*(x + y - 1)/2, (y - 1)/x/y + 1/y]

In [12]: eqs = [eq.as_numer_denom()[0].factor() for eq in eqs]

In [13]: eqs
Out[13]: [-y⋅(x + y - 1), y⋅(x + y - 1)]
```
Hence if `x + y - 1 = 0` then the system is solved. The exceptions are that we should have `y != 0` and `x != 0` to avoid divide by zero. In other words provided `y not in {0, 1}` any `(1-y,y)` is a solution. Usually `solve` does not take account of degenerate cases like this though so the result `(1 - y, y)` is expected.

oscarbenjamin commented:
> the result `(1 - y, y)` is expected.

At the same time I don't consider a failure to produce that result to be a blocker. Since `solve` is already unreliable any good change tends to make some right results wrong and some wrong results right.

asmeurer committed:
Update a doctest in _invert

asmeurer committed:
Update a solver test output

See the discussion on #24553.

asmeurer commented:
I've updated the two failing tests to give the new output. If we are OK with that new output, then this should be ready to go.

oscarbenjamin commented:
Should there not be a test for the issues that are fixed?

asmeurer commented:
Oops, I forgot to copy the test over from the other PR.

asmeurer committed:
Add a test for issue #24368

asmeurer commented:
I didn't notice the slow test failure here:

```py
a, b, x = symbols('a b x')
ax = a**(3*x + 5)
ans = solve(3*log(ax) + b*log(ax) + ax, x)
x0 = 1/log(a)
x1 = sqrt(3)*I
x2 = b + 3
x3 = x2*LambertW(1/x2)/a**5
x4 = x3**Rational(1, 3)/2
assert ans == [
    x0*log(x4*(-x1 - 1)),
    x0*log(x4*(x1 - 1)),
    x0*log(x3)/3]
```

With this PR it now only gives:

```py
>>> ans
[-5/3 - LambertW(1/(b + 3))/(3*log(a))]
```

asmeurer commented:
According to `equals`, `-5/3 - LambertW(1/(b + 3))/(3*log(a))` is not a correct solution. So something has gone wrong here.

asmeurer commented:
If we make `_invert` look a little more like it used to

```diff
diff --git a/sympy/solvers/solvers.py b/sympy/solvers/solvers.py
index 403656ba03..5abc16cf1c 100644
--- a/sympy/solvers/solvers.py
+++ b/sympy/solvers/solvers.py
@@ -3163,8 +3163,11 @@ def _invert(eq, *symbols, **kwargs):
                 if not isinstance(simp, Pow):
                     ad = exp(a_exp*log(a_base))
                     bd = exp(b_exp*log(b_base))
-                lhs = powsimp(powdenest(ad/bd))
-                rhs = -bi/ai
+                rat = ad/bd
+                _lhs = powsimp(rat)
+                if _lhs != rat:
+                    lhs = _lhs
+                    rhs = -bi/ai
             elif ai == -bi:
                 if isinstance(ad, Function) and ad.func == bd.func:
                     if len(ad.args) == len(bd.args) == 1:
```

the original issue is still fixed (and the check=False test discussed above still gives the same thing), and the LambertW test now gives

```
>>> a, b = symbols('a b')
>>> ax = a**(3*x + 5)
>>> eq = 3*log(ax) + b*log(ax) + ax
>>> ans = solve(eq, x)
>>> ans
[(-(b + 3)*LambertW((-1 - sqrt(3)*I)/(2*(b + 3))) - log(a**(5*b + 15)))/(3*(b + 3)*log(a)), (-(b + 3)*LambertW((-1 + sqrt(3)*I)/(2*(b + 3))) - log(a**(5*b + 15)))/(3*(b + 3)*log(a)), (-(b + 3)*LambertW(1/(b + 3)) - log(a**(5*b + 15)))/(3*(b + 3)*log(a))]
>>> pprint(ans)
⎡           ⎛-1 - √3⋅ⅈ⎞      ⎛ 5⋅b + 15⎞             ⎛-1 + √3⋅ⅈ⎞      ⎛ 5⋅b + 15⎞             ⎛  1  ⎞      ⎛ 5⋅b + 15⎞⎤
⎢- (b + 3)⋅W⎜─────────⎟ - log⎝a        ⎠  - (b + 3)⋅W⎜─────────⎟ - log⎝a        ⎠  - (b + 3)⋅W⎜─────⎟ - log⎝a        ⎠⎥
⎢           ⎝2⋅(b + 3)⎠                              ⎝2⋅(b + 3)⎠                              ⎝b + 3⎠                 ⎥
⎢───────────────────────────────────────, ───────────────────────────────────────, ───────────────────────────────────⎥
⎣            3⋅(b + 3)⋅log(a)                         3⋅(b + 3)⋅log(a)                       3⋅(b + 3)⋅log(a)         ⎦
```

Here's the expected answer from the test

```py
>>> expected
[log(((b + 3)*LambertW(1/(b + 3))/a**5)**(1/3)*(-1 - sqrt(3)*I)/2)/log(a), log(((b + 3)*LambertW(1/(b + 3))/a**5)**(1/3)*(-1 + sqrt(3)*I)/2)/log(a), log((b + 3)*LambertW(1/(b + 3))/a**5)/(3*log(a))]
>>> pprint(expected)
⎡   ⎛       __________________            ⎞     ⎛       __________________            ⎞                       ⎤
⎢   ⎜      ╱          ⎛  1  ⎞             ⎟     ⎜      ╱          ⎛  1  ⎞             ⎟                       ⎥
⎢   ⎜     ╱  (b + 3)⋅W⎜─────⎟             ⎟     ⎜     ╱  (b + 3)⋅W⎜─────⎟             ⎟                       ⎥
⎢   ⎜    ╱            ⎝b + 3⎠             ⎟     ⎜    ╱            ⎝b + 3⎠             ⎟     ⎛         ⎛  1  ⎞⎞⎥
⎢   ⎜   ╱    ──────────────── ⋅(-1 - √3⋅ⅈ)⎟     ⎜   ╱    ──────────────── ⋅(-1 + √3⋅ⅈ)⎟     ⎜(b + 3)⋅W⎜─────⎟⎟⎥
⎢   ⎜3 ╱             5                    ⎟     ⎜3 ╱             5                    ⎟     ⎜         ⎝b + 3⎠⎟⎥
⎢   ⎜╲╱             a                     ⎟     ⎜╲╱             a                     ⎟  log⎜────────────────⎟⎥
⎢log⎜─────────────────────────────────────⎟  log⎜─────────────────────────────────────⎟     ⎜        5       ⎟⎥
⎢   ⎝                  2                  ⎠     ⎝                  2                  ⎠     ⎝       a        ⎠⎥
⎢──────────────────────────────────────────, ──────────────────────────────────────────, ─────────────────────⎥
⎣                  log(a)                                      log(a)                           3⋅log(a)      ⎦
```

Numerically these seem equal. 

It's maybe a little concerning that the previous change in `_invert` was leading to the apparently wrong answer. It seems mathematically valid, so I wonder if some other code in the lambertw solver is incorrect somehow.

oscarbenjamin commented:
> I wonder if some other code in the lambertw solver is incorrect somehow.

I'm sure it is. In `solve` routines are unreliable enough that fixing a bug can break something and breaking something can make some wrong result become right.

asmeurer commented:
I think the two solutions are the same modulo some assumptions. According to https://en.wikipedia.org/wiki/Lambert_W_function#Identities, $\log(W(x)) = \log(x) - W(x)$ for $x > 0$. It's applied this via implicitly using $\log(xy) = \log(x) + \log(y)$. It's also converted $(b + 3)\log\left(a^5\right)$ into $\log\left(a^{5b + 15}\right)$ (which otherwise ought to be cancelled). 

So I'm not sure. Numerically the answers seem to be the same, at least according to `equals`. They are symbolically equal, but only after applying log and Lambert W identities that aren't equal for all x. 

And honestly, trying to simplify either answer in the original equation is also challenging without using these log and Lambert W identities that aren't true for their whole domains.

asmeurer commented:
And actually `-5/3 - LambertW(1/(b + 3))/(3*log(a))` is also the same as after applying some of these identities 

```py
>>> expand(cancel((expand_log(ans[-1], force=True))))
-5/3 - LambertW(1/(b + 3))/(3*log(a))
```

I'm not sure how rigorous the Lambert W solver is about these log identities, or where it actually matters in the solution. It's also not clear to me why `equals` thinks this solution isn't equal, since some random numerical substitutions for `a` and `b` seem to give near 0:

```py
>>> eq = 3*log(ax) + b*log(ax) + ax
>>> ans2 = S("-5/3 - LambertW(1/(b + 3))/(3*log(a))")
>>> eq.subs(x, ans2).equals(0)
False
>>> eq.subs(x, ans2).subs({a: -1.0, b: -5.0}).evalf()
-1.3349922906734e-16 - 1.44409343519585e-16*I
>>> eq.subs(x, ans2).subs({a: 6.0, b: 5.0}).evalf()
-2.22044604925031e-16
```

If anyone more versed in Lambert W wants to play around with these various solutions in the original equation:

```py
>>> pprint(eq)
 3⋅x + 5        ⎛ 3⋅x + 5⎞        ⎛ 3⋅x + 5⎞
a        + b⋅log⎝a       ⎠ + 3⋅log⎝a       ⎠
```

and see which seem mathematically valid for all `a` and `b`, that would be nice. Otherwise, it seems to me like the new solutions here are still fine (maybe a little less simple, but that seems like a problem for the solver itself).

asmeurer commented:
Well it seems that there might be more to do after all, because the next test in that function still gives just one solution now instead of three:

```py
x1 = LambertW(Rational(1, 3))
x2 = a**(-5)
x3 = -3**Rational(1, 3)
x4 = 3**Rational(5, 6)*I
x5 = x1**Rational(1, 3)*x2**Rational(1, 3)/2
ans = solve(3*log(ax) + ax, x)
assert ans == [
    x0*log(3*x1*x2)/3,
    x0*log(x5*(x3 - x4)),
    x0*log(x5*(x3 + x4))]
```

```py
>>> solve(3*log(ax) + ax, x)
[-5/3 - LambertW(1/3)/(3*log(a))]
```

asmeurer commented:
Digging into `_invert` for that equation in master vs. this branch, it does really seem to boil down to that solve actually does benefit from rewriting powers into exponentials. When stepping through `_invert` in a debugger in master, one sees internal expressions like

```py
 5  3⋅x⋅log(a)        ⎛ 5  3⋅x⋅log(a)⎞
a ⋅ℯ           + 3⋅log⎝a ⋅ℯ          ⎠
```

The code in this PR wouldn't create an expression like that because it only looks at if the top-level expression is a power, not inner expressions too. 

So I'm beginning to wonder if https://github.com/sympy/sympy/pull/24549 actually is a better fix here after all.

asmeurer commented:
I think it has something to do with this:

```py
>>> u = Dummy('rhs')
>>> a, x, y = symbols('a x y')
>>> solve(log(a**(3*x + 5)) - y, x)
[log(exp(y)/a**5)/(3*log(a)), log((exp(y)/a**5)**(1/3)*(-1 - sqrt(3)*I)/2)/log(a), log((exp(y)/a**5)**(1/3)*(-1 + sqrt(3)*I)/2)/log(a)]
>>> solve(log(a**(3*x + 5)) - u, x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/solvers/solvers.py", line 1144, in solve
    solution = _solve(f[0], *symbols, **flags)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/solvers/solvers.py", line 1690, in _solve
    raise NotImplementedError('\n'.join([msg, not_impl_msg % f]))
NotImplementedError:
No algorithms are implemented to solve equation -_rhs + log(a**(3*x + 5))
```

When I internally replace this Dummy (in bivariate.py) with a symbol, I get the full set of answers.

asmeurer commented:
It's because of https://github.com/sympy/sympy/blob/242c87f7b55da8170ed46f5a9d36514e67f6e39b/sympy/solvers/solvers.py#L2649-L2650

I don't know what that is actually supposed to filter, but it prevents recursive solving from working here. @smichr any thoughts? The recursive solve is this one https://github.com/sympy/sympy/blob/242c87f7b55da8170ed46f5a9d36514e67f6e39b/sympy/solvers/bivariate.py#L153-L154

asmeurer committed:
Make _invert work more like it used to

This makes some tests still work similar to how they did before, but still
leaves #24368 fixed.

However, it does change the behavior of some of the slow Lambert W tests, so
this approach may still require some thought.

asmeurer commented:
These are the tests that fail if I comment out that `is_Dummy` line (in master):

```
__________________________________________________________ sympy/solvers/tests/test_solvers.py:test_issue_4463 __________________________________________________________
Traceback (most recent call last):
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/solvers/tests/test_solvers.py", line 1396, in test_issue_4463
    assert solve(((x - 3)*(x - 2))**((x - 3)*(x - 4))) == [2]
AssertionError
_________________________________________________________________________________________________________________________________________________________________________
_________________________________________________________ sympy/solvers/tests/test_solvers.py:test_base_0_exp_0 _________________________________________________________
Traceback (most recent call last):
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/solvers/tests/test_solvers.py", line 2145, in test_base_0_exp_0
    assert solve(0**x - 1) == [0]
AssertionError
_________________________________________________________________________________________________________________________________________________________________________
_________________________________________________________ sympy/solvers/tests/test_solvers.py:test_issue_17452 __________________________________________________________
Traceback (most recent call last):
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/solvers/tests/test_solvers.py", line 2426, in test_issue_17452
    assert solve(x**(x/11) + pi/11, x) == [exp(LambertW(-11*log(11) + 11*log(pi) + 11*I*pi))]
AssertionError
```

The last one just gives a slightly different form of the answer, but the first two end up returning `[]` instead of the solution.

asmeurer commented:
> These are the tests that fail if I comment out that is_Dummy line (in master):

Also this slow test:

```
______________________________________________________ sympy/solvers/tests/test_solvers.py:test_lambert_bivariate _______________________________________________________
Traceback (most recent call last):
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/testing/pytest.py", line 188, in func_wrapper
    func()
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/solvers/tests/test_solvers.py", line 1966, in test_lambert_bivariate
    assert solve((x**3)**(x/2) + pi/2, x) == [
AssertionError
_________________________________________________________________________________________________________________________________________________________________________
_________________________________________________________ sympy/solvers/tests/test_solvers.py:test_issue_15731 __________________________________________________________
Traceback (most recent call last):
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/testing/pytest.py", line 188, in func_wrapper
    func()
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/sympy/solvers/tests/test_solvers.py", line 2365, in test_issue_15731
    assert solve(Eq((x**2 - 7*x + 11)**(x**2 - 13*x + 42), 1)) == [2, 3, 4, 5, 6, 7]
AssertionError
```

The first one just gives a different form of the answer, but the second one gives `[]`.

It seems odd to me that removing a `raise NotImplementedError` causes solve to return fewer solutions.

asmeurer commented:
If I disable the Dummy check, I get a rather weird test failure:

```py
>>> solve(0**x - 1)
[]
```

(normally it should return `[0]`).

The reason this is happens is because internally, in a recursive solve call, it solves `0**x - t`, where `t` is a Dummy. This produces the result `0`, but the check removes the solution because 1 != t. This has me confused. 

- Why does solve call `check` in recursive calls? Shouldn't it only be necessary to check the final solutions at the end.

- Why does solve turn `x**0 - 1` into `x**0 - t`? This, along with the check produces a solution that is wrong, because it's only correct when t = 1.

Another curious test failure is this (from test_issue_15731):

```py
assert solve(x**(2/x) - 2) == [2, 4]
```

It now gives:

```py
>>> solve(x**(2/x) - 2)
[2, -LambertW(-log(sqrt(2)) - I*pi)/(log(sqrt(2)) + I*pi)]
```

The LambertW solution is nonreal, and as far as I can tell, correct. But it also misses the `4` solution it found before. I haven't investigated it in depth yet, but my guess it is takes a different soling code path which makes it find a different solution set. 

I'm unclear if there's a simple generic fix here that doesn't break other things. The whole thing feels fragile, but it's better to keep existing things working than to fix something that doesn't work. 

Here's the diff I'm using to test this:

```diff
diff --git a/sympy/solvers/solvers.py b/sympy/solvers/solvers.py
index fcc7e13c31..5e6f7107f1 100644
--- a/sympy/solvers/solvers.py
+++ b/sympy/solvers/solvers.py
@@ -2667,9 +2667,9 @@ def _tsolve(eq, sym, **flags):
             if sym not in lhs.exp.free_symbols:
                 return _vsolve(lhs.base - rhs**(1/lhs.exp), sym, **flags)

-            # _tsolve calls this with Dummy before passing the actual number in.
-            if any(t.is_Dummy for t in rhs.free_symbols):
-                raise NotImplementedError # _tsolve will call here again...
+            # # _tsolve calls this with Dummy before passing the actual number in.
+            # if any(t.is_Dummy for t in rhs.free_symbols):
+            #     raise NotImplementedError # _tsolve will call here again...

             # a ** g(x) == 0
             if not rhs:
@@ -2757,6 +2757,8 @@ def equal(expr1, expr2):
                         ok = eq.subs(sym, s).equals(0)
                     if ok:
                         sol.append(s)
+                if not sol:
+                    raise NotImplementedError
                 return list(set(sol))

         elif lhs.is_Function and len(lhs.args) == 1:
```

I'm still feeling that https://github.com/sympy/sympy/pull/24549 is a better fix for the original issue. Unless someone can suggest a straightforward fix that doesn't involve totally refactoring the way `solve` works (which clearly needs to happen, but it way out of scope for what I'm trying to do here).

oscarbenjamin commented:
> * Why does solve call `check` in recursive calls? Shouldn't it only be necessary to check the final solutions at the end.

There is a benefit in calling check in recursive calls because it saves any further processing of invalid solutions. The way that the checking is handled in solve is very broken though. The equation used for checking should be the original equation before any rewrites or other changes else but that equation doesn't even make it through to `_solve`.

In a rewrite of solve I wouldn't have the checking at all but rather would encode the conditions that are supposed to be checked as new variables. The need for checking comes from transformations such as `unrad` that introduce spurious solutions. The approach used in solve though is bad because it forgets what conditions need to be checked and then later tries to just check by substituting the solution into the equation but that's a really hard way to check things and does not generally work for symbolic solutions. It's much better to remember the condition that is supposed to be checked.

For denominators just introduce a new variable so given:
```python
In [51]: eq = (x - 1)/x + 1/x

In [52]: eq
Out[52]: 
x - 1   1
───── + ─
  x     x
```
The solution is to find the denominator and make it equal to a new variable `d` and require `d` to be nonzero. Then the equations to be solved are:
```python
In [53]: d = symbols('d')

In [54]: num, den = eq.as_numer_denom()

In [55]: eq1 = num

In [56]: eq2 = den - d

In [57]: eq1, eq2
Out[57]: (x, -d + x)

In [58]: solve([eq1, eq2], [x, d])
Out[58]: {d: 0, x: 0}
```
But now we know that `d != 0` is needed. No need for complicated checking. Had the result for `d` been symbolic we could have used it to make a symbolc condition for the validity of the solution.

The same thing goes for `unrad`. It does things like square both sides of an equation to eliminate radicals but that introduces spurious solutions. Better to introduce new variables so if you have:
```python
In [59]: eq = sqrt(x + 1) - x

In [60]: eq
Out[60]: 
       _______
-x + ╲╱ x + 1 
```
Then introduce two new variables `r` and `p` with the equations:
```python
In [70]: p, r = symbols('p, r')

In [71]: pval = x + 1

In [72]: rval = sqrt(pval)

In [73]: eq1 = eq.subs(rval, r)

In [74]: eq2 = p - pval

In [75]: eq3 = r**2 - p

In [76]: eqs = [eq1, eq2, eq3]

In [77]: eqs
Out[77]: 
⎡                        2⎤
⎣r - x, p - x - 1, -p + r ⎦

In [78]: groebner(eqs, [r, p, x])
Out[78]: 
             ⎛⎡                   2        ⎤                              ⎞
GroebnerBasis⎝⎣r - x, p - x - 1, x  - x - 1⎦, r, p, x, domain=ℤ, order=lex⎠
```
Now the solutions are the roots of `x**2 - x - 1` but only if substituting that into the other equations gives a result satisfying `sqrt(p) = r` which again is a condition that can easily be checked numerically or otherwise represetned symbolically.

oscarbenjamin commented:
> The whole thing feels fragile

That's the problem with working on anything here. So many things only just about work that any slight change breaks things so it isn't possible to fix any one thing without breaking something else.

It is unfortunate that the desire to rewrite solve went into solveset which is really aimed at a very different problem. What would have been more useful for users would literally be to just rewrite solve even if it's just called solve2 but made consistent and reliable and built with solid foundations.

asmeurer commented:
I think the general principle of solveset is fine, but more work needs to be put into simplifying set expressions, and reducing them into solve() like outputs. There's no point in having two different solvers, when they do the same thing. But the most general solver has to be something like solveset because in general solutions can only be represented by infinite sets.

The design in solveset is definitely cleaner than solve, but still doesn't quite look like what I think it should. The dsolve design is a lot closer to what I'd like to se, with different hints for different solvers. What we need is that, plus a better abstraction on rewriting, and a better abstraction for recursive solving. Rewriting is also missing from dsolve, but it's not a big deal there. Recursive solving usually doesn't happen in dsolve, although the solvers that do attempt to do it are already problematic because not enough thought has been put into them.

All the other stuff that happens in solve should happen as completely separate pre- or post- processing steps. 

> But now we know that d != 0 is needed. No need for complicated checking. Had the result for d been symbolic we could have used it to make a symbolc condition for the validity of the solution.

Agree with this. If solve did this right it could just return Piecewise solutions. 

This is basically what I was trying to get at with https://github.com/sympy/sympy/issues/24650#issuecomment-1435472628. With something like `Eq(1/x, x + 1/x)` instead of carefully trying to maintain everything so that check works, just separately note that the `x`s are cancelled for the check stage at the end. In general, you just need to make a note of the denominators in the expression once upfront and just check them directly at the end. 

For unrad, I agree that checking should just be part of its own internal routine rather than done at the end.

asmeurer commented:
I've no confidence that I can ever make this work. I agree that it's in principle the better solution, but it just breaks too many things. Without a full rewrite of solve in the ways I suggested above, I don't really feel this work is worth the time. I'm closing this in favor of https://github.com/sympy/sympy/pull/24999, which fixes the original issue in a much simpler way.
